blockchain/vm/evm.go

```diff
@@ -538,6 +538,35 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
 	return evm.create(caller, codeAndHash, gas, value, contractAddr, humanReadable, codeFormat)
 }
 
+func (evm *EVM) GetPrecompiledContractMap(caller common.Address) map[common.Address]PrecompiledContract {
+	if codeInfo, ok := evm.StateDB.GetCodeInfo(caller); ok {
+		var (
+			currentBlockVMVersion = params.GenerateVmVersion(evm.chainRules)
+			contractVmVersion     = codeInfo.GetVmVersion()
+		)
+
+		// There are contracts which uses the old precompiled contract map (use the map at deployment time)
+		//      (gas price policy also follows old map's rule)
+		if contractVmVersion < currentBlockVMVersion {
+			// If new "VmVersion of CodeInfo" is added, please add new case below
+			switch contractVmVersion {
+			default:
+				// Without this version, contracts that are deployed before istanbul and uses 0x09-0x0b won't work properly.
+				return PrecompiledContractsConstantinople
+			}
+		}
+	}
+
+	// There are contracts which uses latest precompiled contract map (regardless of deployment time)
+	// If new HF is added, please add new case below
+	switch {
+	case evm.chainRules.IsIstanbul:
+		return PrecompiledContractsIstanbul
```

Comment:
Shouldn't we add `IsConstantinople` in Rules?

Reply:
@winnie-byun Can I ask why this field is needed? In here, default case means use PrecompiledContractsConstantinople, not a Constantinople hard fork.

Reply:
For the next vm version, we have to make `IsConstantinople`. It seems awkward to me if we make something for the previous version.
It is not critical problem.

Reply:
`IsConstantinople` is not needed. If new vm version is added, add a new case statement as shown in the code below.
```
switch contractVmVersion {
case VmVersionIstanbul:
        return PrecompiledContractsBerlin (last hard fork of that vm version)
default:
        return PrecompiledContractsConstantinople
```